src/main/java/uk/gov/hmcts/reform/prl/clients/ccd/CaseAssignmentService.java

```diff
@@ -0,0 +1,426 @@
+package uk.gov.hmcts.reform.prl.clients.ccd;
+
+import com.fasterxml.jackson.databind.ObjectMapper;
+import feign.FeignException;
+import lombok.Builder;
+import lombok.RequiredArgsConstructor;
+import lombok.extern.slf4j.Slf4j;
+import org.springframework.stereotype.Service;
+import uk.gov.hmcts.reform.authorisation.generators.AuthTokenGenerator;
+import uk.gov.hmcts.reform.ccd.client.CaseAssignmentApi;
+import uk.gov.hmcts.reform.ccd.client.model.CaseAssignmentUserRoleWithOrganisation;
+import uk.gov.hmcts.reform.ccd.client.model.CaseAssignmentUserRolesRequest;
+import uk.gov.hmcts.reform.prl.enums.noticeofchange.BarristerRole;
+import uk.gov.hmcts.reform.prl.enums.noticeofchange.BarristerRole.Representing;
+import uk.gov.hmcts.reform.prl.exception.GrantCaseAccessException;
+import uk.gov.hmcts.reform.prl.models.Element;
+import uk.gov.hmcts.reform.prl.models.OrgSolicitors;
+import uk.gov.hmcts.reform.prl.models.complextypes.PartyDetails;
+import uk.gov.hmcts.reform.prl.models.dto.barrister.AllocatedBarrister;
+import uk.gov.hmcts.reform.prl.models.dto.ccd.Barrister;
+import uk.gov.hmcts.reform.prl.models.dto.ccd.CaseData;
+import uk.gov.hmcts.reform.prl.models.roleassignment.getroleassignment.RoleAssignmentResponse;
+import uk.gov.hmcts.reform.prl.models.roleassignment.getroleassignment.RoleAssignmentServiceResponse;
+import uk.gov.hmcts.reform.prl.services.OrganisationService;
+import uk.gov.hmcts.reform.prl.services.RoleAssignmentService;
+import uk.gov.hmcts.reform.prl.services.SystemUserService;
+import uk.gov.hmcts.reform.prl.utils.MaskEmail;
+
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.Set;
+import java.util.UUID;
+import java.util.function.Supplier;
+import java.util.stream.Stream;
+
+import static java.util.stream.Collectors.collectingAndThen;
+import static java.util.stream.Collectors.toList;
+import static uk.gov.hmcts.reform.prl.constants.PrlAppsConstants.C100_CASE_TYPE;
+import static uk.gov.hmcts.reform.prl.constants.PrlAppsConstants.FL401_CASE_TYPE;
+import static uk.gov.hmcts.reform.prl.enums.noticeofchange.BarristerRole.Representing.DAAPPLICANT;
+import static uk.gov.hmcts.reform.prl.enums.noticeofchange.BarristerRole.Representing.DARESPONDENT;
+
+@Slf4j
+@Builder
+@RequiredArgsConstructor
+@Service
+public class CaseAssignmentService {
+    private final CaseAssignmentApi caseAssignmentApi;
+    private final SystemUserService systemUserService;
+    private final AuthTokenGenerator tokenGenerator;
+    private final OrganisationService organisationService;
+    private final RoleAssignmentService roleAssignmentService;
+    private final MaskEmail maskEmail;
+    private final ObjectMapper objectMapper;
+
+    private static IllegalArgumentException getIllegalArgumentException(CaseData caseData,
```

Comment:
I think it would be nice to have a separate exception class of it

Reply:
Replaced with InvalidPartyIdException